### DISCUSSION TITLE
RFC: Operation datastore

### DISCUSSION THREAD

**Discussion by damip** [massalabs/massa#2648] [Ideas]
# RFC: Operation datastore

Currently, to deploy smart contracts, we need to store their bytecode inside another smart contract bytecode which might be suboptimal.

Same for websites etc... and arbitrary data, used for example to initialize the datastore of an SC being created.

We propose adding a `op_datastore: HashMap<Hash, Vec<u8>>` field to the `ExecuteSC` and `CallSC` operations, allowing the caller to inject data in the operation that can be accessed by the executed smart contract through a specialized `op_data_get(key) -> Vec<u8>` ABI

Here are some use cases:
* avoid the need for transformers etc... to inject bytes into the code itself
  * make SC deployment pipeline cleaner (no more transformer version incompatibilities etc... and 1 less compilation step)
  * allow having a pre-compiled simple SC deployer that doesn't need to be compiled when the SC to deploy changes
  * reduce the wasmer load on parsing large SCs containing mostly bytes
* have pre-compiled SCs allowing you to deploy SCs, write in ledger datastores, etc... just by changing the op datastore data being sent and without needing to recompile the SC itself. A library of such precompiled SCs can then be built and behave a bit like the specialized CallSC but for more varied purposes including:
  * deploy SCs
  * atomically call multiple functions in a row (like an atomic all-or-nothing multi-CallSC)
  * write data into the ledger datastore
  * => this allows executing advanced instructions (beyond CallSC) from a web3 page without the need to inline a compiler
* allow injecting arbitrary data into the blocks only and not into the ledger
  * this allows outside layer-2 observers software to catch the data, execute bytecode off-chain, store data off-chain, trigger real-world actions etc...

**Comment by AurelienFT** [comment]
I don't understand what does it change having it in a datastore or in the bytecode directly.

I think something good that could be made would be to improve tooling so that we can use a CLI to upload our SC or website without having to create a wrapper SC and be more similar with other blockchains.

**Reply by damip** [reply]
added some use case examples

**Comment by damip** [comment]
Multiple Massa partners have expressed the need.for this feature. Prioritizing it soon.
